### ISSUE TITLE
COPY / The system cannot find the file specified.

### ISSUE THREAD

**Issue by StarkovArt** [moby/moby#34893]
I can't build image 
Always this error, and if I call docker-compose same.

```
docker build -t tagname .

Sending build context to Docker daemon  3.584kB
Step 1/6 : FROM microsoft/aspnetcore:2.0
 ---> 7789f525cb73
Step 2/6 : ARG source
 ---> Using cache
 ---> 9072545d21be
Step 3/6 : WORKDIR /app
 ---> Using cache
 ---> 1d8395c9a508
Step 4/6 : EXPOSE 80
 ---> Using cache
 ---> 748b65344910
**Step 5/6 : COPY ${source:-obj/Docker/publish} .
COPY failed: GetFileAttributesEx \\?\C:\Windows\TEMP\docker-builder460102490\obj\Docker\publish: The system cannot find the file specified.**
```

```
DockerFile
FROM microsoft/aspnetcore:2.0
ARG source
WORKDIR /app
EXPOSE 80
COPY ${source:-obj/Docker/publish} .
ENTRYPOINT ["dotnet", "docker-example.dll"]
```

```
docker version
Client:
 Version:      17.06.2-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 19:57:19 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.06.2-ce
 API version:  1.30 (minimum version 1.24)
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 19:59:47 2017
 OS/Arch:      windows/amd64
 Experimental: true
```

docker version

```
Client:
 Version:      17.06.2-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 19:57:19 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.06.2-ce
 API version:  1.30 (minimum version 1.24)
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 19:59:47 2017
 OS/Arch:      windows/amd64
 Experimental: true
```

```
D:\docker-example\docker-example>docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 5
Server Version: 17.06.2-ce
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd json-file logentries splunk syslog
Swarm: inactive
Default Isolation: hyperv
Kernel Version: 10.0 15063 (15063.0.amd64fre.rs2_release.170317-1834)
Operating System: Windows 10 Pro
OSType: windows
Architecture: x86_64
CPUs: 12
Total Memory: 15.95GiB
Name: A-STARKOU
ID: 5LPB:CED6:FAN5:OYGT:6GTO:ZA3T:F2XQ:DNIS:KAIJ:LXCI:S26L:DB3N
Docker Root Dir: C:\ProgramData\Docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: -1
 Goroutines: 24
 System Time: 2017-09-19T17:39:08.3799151+03:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
Silly question, but;

- is the file (or directory) `obj/Docker/publish`  present in the directory that you're building from (probably `D:\docker-example\docker-example`)?
- is there a `.dockerignore` file in your build directory?

**Comment by StarkovArt** [comment]
- 1. D:\docker-example\docker-example сorrect directory
- 2. Yes,  *!obj/Docker/publish/ *!obj/Docker/empty/

**Comment by thaJeztah** [comment]
Does it work if you temporarily remove / rename the `.dockerignore` file?

**Comment by StarkovArt** [comment]
Remove .dockerignore file 
D:\docker-example\docker-example>docker build -t firs-example .
Sending build context to Docker daemon   8.24MB
Step 1/6 : FROM microsoft/aspnetcore:2.0
 ---> 0983233db220
Step 2/6 : ARG source
 ---> Using cache
 ---> e6e3fdae1f43
Step 3/6 : WORKDIR /app
 ---> Using cache
 ---> 27d1cc891d3c
Step 4/6 : EXPOSE 80
 ---> Using cache
 ---> b2a3262c9fd4
Step 5/6 : COPY ${source:-obj/Docker/publish} .
COPY failed: GetFileAttributesEx \\?\C:\Windows\TEMP\docker-builder216221870\obj\Docker\publish: The system cannot find the file specified.

**Comment by StarkovArt** [comment]
I don't have error only when I use dotnet publish ./Microservice.Organization.sln -c Release -o ./obj/Docker/publish and  docker build -t test-app it's work

**Comment by thaJeztah** [comment]
ping @friism anything you can help here?

**Comment by friism** [comment]
It looks like this file was generated using the Visual Studio tools for Docker. These files are not very good and will only work when invoked from within Visual Studio (or you can figure out the right way to pass what I think is the `source` parameter.

For a better approach, check out these resources:

 * https://docs.docker.com/engine/examples/dotnetcore/
 * https://github.com/docker/labs/blob/master/windows/modernize-traditional-apps/modernize-aspnet/part-1.md

**Comment by clawrenceks** [comment]
@friism @StarkovArt I am having the same issue here - I am using the Visual Studio Tools for Docker and I believe the source build argument is required for the debugging inside containers to work correctly. I am however passing in the correct value as the source build argument with docker-compose and I get exactly the same error as provided by the original poster.

**Comment by clawrenceks** [comment]
@friism @StarkovArt It seems to me that the error is caused when the copy instruction is given a location that is abolute. With an absolute path the error in the original post is seen. As long as a path relative to the build context (Dockerfile) is used then in my experience this works as I would expect.

**Comment by PatrickLang** [comment]
I'm seeing the same thing. `.dockerignore` shouldn't be ignoring `obj/Docker/publish` but it still is:

```
*
!obj\Docker\publish\*
!obj\Docker\empty\
```

```
PS C:\Users\patrick\source\repos\fabrikamfiber\FabrikamFiber.Web> dir .\obj\Docker\Publish\


    Directory: C:\Users\patrick\source\repos\fabrikamfiber\FabrikamFiber.Web\obj\Docker\Publish


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----        11/9/2017  12:10 AM                bin
d-----        11/9/2017  12:10 AM                Content
d-----        11/9/2017  12:10 AM                Scripts
d-----        11/9/2017  12:10 AM                Views
-a----        11/8/2017  10:18 PM            109 Global.asax
-a----        11/8/2017  10:18 PM            959 packages.config
-a----        11/9/2017  12:10 AM           6184 Web.config
```


Notice that the build context is 3.072kB - it's definitely not including any binaries

```
PS C:\Users\patrick\source\repos\fabrikamfiber\FabrikamFiber.Web> docker build -t patricklang/fabrikamfiber.web:1709 .
Sending build context to Docker daemon  3.072kB
Step 1/4 : FROM microsoft/aspnet:4.7.1-windowsservercore-1709
 ---> ac5761dc8335
Step 2/4 : WORKDIR /inetpub/wwwroot
 ---> Using cache
 ---> 17e823a41cfb
Step 3/4 : RUN c:\windows\system32\inetsrv\appcmd.exe set AppPool DefaultAppPool '-processModel.identityType:LocalSystem'
 ---> Using cache
 ---> ee2b99f6d09d
Step 4/4 : COPY obj/Docker/publish .
COPY failed: GetFileAttributesEx \\?\C:\Users\patrick\AppData\Local\Temp\2\docker-builder625515923\obj\Docker\publish: The system cannot find the path specified.
```

**Comment by PatrickLang** [comment]
I also tried a `%s/\\/\//g` on the .dockerignore and although it looks prettier it didn't change the behavior

**Comment by clawrenceks** [comment]
@PatrickLang Could you include the content of your Dockerfile please?

**Comment by PatrickLang** [comment]
Dockerfile:

```
FROM microsoft/aspnet:4.7.1-windowsservercore-1709
WORKDIR /inetpub/wwwroot
RUN c:\windows\system32\inetsrv\appcmd.exe set AppPool 
COPY obj/Docker/publish .
```

**Comment by PatrickLang** [comment]
My dockerfile originally had `COPY ${source:-obj/Docker/publish}` but I changed it to the above and removed the `ARG`. It failed the same way before and after that change

**Comment by ShaunPseudocode** [comment]
I'm going thru the same problem @PatrickLang

**Comment by thaJeztah** [comment]
Trying to reproduce this issue, but on a Linux machine, so definitely not the same situation.

Thinking of possible issues with the `.dockerignore` (we have run into some corner-cases in the past), such as;

- `!obj/Docker/publish/*` evaluating the last wildcard (`*`), and therefore matching _subdirectories_ of `obj/Docker/publish/`, but not `obj/Docker/publish/` itself
- no leading slash / backslash in the `.dockerignore` rules (https://github.com/moby/moby/pull/32088)
- same but for trailing slash/backslash (possibly leading it to be treated as a _prefix_ instead of a directory/path)
- the `WORKDIR` statement interfering with matching (don't think we had issues with that in the past though)


I wrote down a script that should reproduce the problem being discussed here. It's written with the assumption of a Linux machine / bash, so it would be appreciated if someone running on Windows could adopt these steps to a Windows equivalent, and try to reproduce:



Change to your home-directory; different drives (non-`C:\`), or paths outside of home (at least on Docker for Mac/Windows) could potentially be an issue, so excluding that possibility:

```bash
cd ~/
```

Create a fresh directory for this test; to make sure nothing else is interfering with the test :smile:. Inside that directory, create subdirectories similar to the ones being discussed here, as well as some files (for testing), and paths _not_ included in the `.dockerignore`.

```bash
mkdir repro-34893 && cd repro-34893

mkdir -p obj/Docker/empty
mkdir -p obj/Docker/empty/sub1
mkdir -p obj/Docker/empty/sub2

mkdir -p obj/Docker/publish
mkdir -p obj/Docker/publish/sub1
mkdir -p obj/Docker/publish/sub2

mkdir -p something
mkdir -p something/sub1
mkdir -p something/sub2


touch root1-should-be-excluded
touch something/empty1-should-be-excluded
touch something/sub1/empty-sub1-should-be-excluded
touch something/sub2/empty-sub2-should-be-excluded

touch obj/Docker/empty/empty1-should-be-there
touch obj/Docker/empty/sub1/empty-sub1-should-be-there
touch obj/Docker/empty/sub2/empty-sub2-should-be-there

touch obj/Docker/publish/publish1-should-be-there
touch obj/Docker/publish/sub1/publish-sub1-should-be-there
touch obj/Docker/publish/sub2/publish-sub2-should-be-there
```

Create the `.dockerignore` file, and a `Dockerfile` that installs `tree` (for debugging the results); it adds both the `obj/Docker/publish` directory to `WORKDIR` (which is the problem being discussed here), but also adds "everything" to a directory named "everything". This allows checking what is, and what _isn't_ excluded;

```bash
cat > .dockerignore -<<'EOF'
*
!obj/Docker/publish/*
!obj/Docker/empty/
EOF

cat > Dockerfile -<<'EOF'
FROM alpine
RUN apk add --no-cache tree
WORKDIR /app
COPY obj/Docker/publish .
RUN tree -a
COPY . /everything
RUN tree -a /everything
EOF
```

The directory on the host should now look something like this:


```bash
tree -a
.
├── .dockerignore
├── Dockerfile
├── obj
│   └── Docker
│       ├── empty
│       │   ├── empty1-should-be-there
│       │   ├── sub1
│       │   │   └── empty-sub1-should-be-there
│       │   └── sub2
│       │       └── empty-sub2-should-be-there
│       └── publish
│           ├── publish1-should-be-there
│           ├── sub1
│           │   └── publish-sub1-should-be-there
│           └── sub2
│               └── publish-sub2-should-be-there
├── root1-should-be-excluded
└── something
    ├── empty1-should-be-excluded
    ├── sub1
    │   └── empty-sub1-should-be-excluded
    └── sub2
        └── empty-sub2-should-be-excluded

11 directories, 12 files
```

Build the image, using the `--no-cache` option to exclude any issue with incorrect matching of the cache:

```
docker build --no-cache -t repro-34893 .
```

On an Ubuntu 16.04 machine, with Docker CE 17.09.0;

```
Sending build context to Docker daemon  8.704kB
Step 1/7 : FROM alpine
 ---> 053cde6e8953
Step 2/7 : RUN apk add --no-cache tree
 ---> Running in 659737601ddf
fetch http://dl-cdn.alpinelinux.org/alpine/v3.6/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.6/community/x86_64/APKINDEX.tar.gz
(1/1) Installing tree (1.7.0-r0)
Executing busybox-1.26.2-r7.trigger
OK: 4 MiB in 12 packages
 ---> 5340fa788584
Removing intermediate container 659737601ddf
Step 3/7 : WORKDIR /app
 ---> f495dfeae00d
Removing intermediate container d65d9e6d9fa0
Step 4/7 : COPY obj/Docker/publish .
 ---> 5bc90ceee00d
Step 5/7 : RUN tree -a
 ---> Running in 3174876adcd9
.
├── publish1-should-be-there
├── sub1
│   └── publish-sub1-should-be-there
└── sub2
    └── publish-sub2-should-be-there

2 directories, 3 files
 ---> 85d5a4da5163
Removing intermediate container 3174876adcd9
Step 6/7 : COPY . /everything
 ---> 7ea9d4b8a5f6
Step 7/7 : RUN tree -a /everything
 ---> Running in 007db9e4d03c
/everything
└── obj
    └── Docker
        ├── empty
        │   ├── empty1-should-be-there
        │   ├── sub1
        │   │   └── empty-sub1-should-be-there
        │   └── sub2
        │       └── empty-sub2-should-be-there
        └── publish
            ├── publish1-should-be-there
            ├── sub1
            │   └── publish-sub1-should-be-there
            └── sub2
                └── publish-sub2-should-be-there

8 directories, 6 files
 ---> b02ae1c061ae
Removing intermediate container 007db9e4d03c
Successfully built b02ae1c061ae
Successfully tagged repro-34893:latest
```

So, on Docker CE 17.09 on Linux, this looks to be working correctly

**Comment by praveenc** [comment]
``COPY`` or ``ADD`` doesn't work (throws the exact same error as mentioned above) - Tried this on both Windows 10 Enterprise and Windows Server 2016 with Docker EE

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 22
Server Version: 17.06.2-ee-6
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd json-file logentries splunk syslog
Swarm: inactive
Default Isolation: process
Kernel Version: 10.0 14393 (14393.1770.amd64fre.rs1_release.170917-1700)
Operating System: Windows Server 2016 Datacenter
OSType: windows
Architecture: x86_64
CPUs: 2
Total Memory: 7GiB
Name: dkwin2k16
ID: V2RF:5WKJ:F2P4:OW5S:I2OP:NOZR:L2B3:OFSL:2QCG:MBCX:LRYX:GYKG
Docker Root Dir: C:\ProgramData\docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```


The issue clearly states that this happens with Docker on windows - Not sure what's the point of trying to reproduce the issue on a Linux machine?

**Comment by thaJeztah** [comment]
@praveenc I commented in https://github.com/Microsoft/iis-docker/issues/60#issuecomment-350931402, but it doesn't look like a bug from that information.

Please keep in mind that the GitHub issue tracker is not intended as a general support forum,
but for [reporting bugs and feature requests](https://raw.githubusercontent.com/moby/moby/master/.github/ISSUE_TEMPLATE.md). For other type of questions, consider using one of;

- the Docker Support Forums - https://forums.docker.com
- the [Docker community Slack channel](https://blog.docker.com/2016/11/introducing-docker-community-directory-docker-community-slack/) (register here: [http://dockr.ly/community](http://dockr.ly/community))
- [StackOverflow](https://stackoverflow.com)

**Comment by abhayongit** [comment]
Hi I have faced this and lost 2 days investigating why COPY and ADD is not working. But I found a solution how COPY and ADD command works in a docker window container. Hope this solution will save some of your time.

1 - ADD and COPY command with source path containing some path other than current folder never works in window container. It should be always
ADD ./filename .
COPY ./filename .
It means copy the image from current directory . If you give any path like
ADD bin/Debug/netcoreapp2.0/publish/ / it always tries to create a directory in \?\C:\ProgramData\Docker\tmp\docker-builder857651318`bin\Debug\netcoreapp2.0\publish`

That means the path you give in source directory will appended and try to create file or folder inside docker random folder (docker-builder857651318) which does not exists. in your case bin\Debug\netcoreapp2.0\publish is always getting appended where nothing exists… So to make COPY and ADD work in docker container do not give any folder path in source path.

2 - Here are steps for clean approach to fix this issue
Create a new folder.
Copy your jar or image in that folder.
Create a Dockerfile in that folder
Give the path like
ADD ./filename .
COPY ./filename .

It means copy from Local folder (./filename) to docker current deployment(.) folder. That works

**Comment by thaJeztah** [comment]
> ADD and COPY command with source path containing some path other than current folder never works in window container.

`ADD` and `COPY` can only use files that are within the build-context; when starting a build, the docker cli compresses the build-context and sends that context to the daemon; after that, the actual build is using those files to build the image.

source _paths_ for `COPY` and `ADD` inside a Dockerfile are also relative to the build-context, so `COPY /some/path` refers to the `some/path` subdirectory inside your build-context.

The build-context is the directory you point at when you start the build, for example (if your current directory is `C:\some-directory`), and you run;

```
docker build -t foobar .
```

The last `.` is indicating that you use the _current_ directory (`C:\some-directory` ) as build-context. Docker now compresses that directory, and send it to the daemon (through the remote API), which then stores the build-context in a temporary directory.

Within the Dockerfile;

```
COPY /some-file.txt /some/target/directory
```

Will copy `some-file.txt` from the root of the build-context (so in this case, a file that was sent from `C:\some-directory\some-file.txt`), and copy it into the image at `/some/target/directory/some-file.txt`.

**Comment by MAkhtarHossain** [comment]
Ok. Let's get to the bottom line. How does the fix look like for the Dockerfile that was referred in beginning of the thread. My build-context folder is c:\abc\xyz. Here is my Dockerfile that I am trying to fix for the same error that I am having at the COPY command:
--------------------------------------------------------------
FROM microsoft/aspnet:4.7.2-windowsservercore-1803
ARG source
WORKDIR /inetpub/wwwroot
COPY ${source:-obj/Docker/publish} .
---------------------------------------------------

**Comment by bearkat2173** [comment]
So I had this same issue and I discovered that Visual Studio only copies files into _obj\Docker\publish_ during Release build.  Visual Studio also uses obj\Docker\publish\ for _source_ parameter when it runs the container (with docker-compose) only when the container is built using the release build; it uses _obj\Docker\empty\_ for Debug build.

The reason I was getting the error is because I had replaced ${source:-obj/Docker/publish} with just obj/Docker/publish.  Once I changed that back my error went away.

If you look at the Docker output in the Output pane of VS, you'll see what VS ends up using for docker-compose configuration.  For Debug builds it creates a volume that points to the files on the host, so changes to files don't require the container to be rebuilt.  For Release builds it copies the files into the container from the _publish_ directory so that the container can be shipped elsewhere.
